gum/gumdarwingrafter.c

```diff
@@ -666,6 +671,222 @@ gum_darwin_grafter_compute_layout (GumDarwinGrafter * self,
   }
 }
 
+static void
+gum_collect_chained_imports (GumDarwinModule * self,
+                             GumCollectImportsOperation * op)
+{
+  GumDarwinModuleImage * image;
+  const GumMachHeader64 * mach_header;
+  gconstpointer command;
+  gsize command_index, segment_index;
+  const GumDarwinSegment * segment, * linkedit_segment;
+
+  if (!gum_darwin_module_ensure_image_loaded (self, NULL))
+    return;
+
+  segment_index = 0;
+  linkedit_segment = NULL;
+
+  while ((segment = gum_darwin_module_get_nth_segment (self, segment_index++))
+      != NULL)
+  {
+    if (strcmp (segment->name, "__LINKEDIT") == 0)
+    {
+      linkedit_segment = segment;
+      break;
+    }
+  }
+
+  if (linkedit_segment == NULL)
+    return;
+
+  image = self->image;
+  mach_header = image->data;
+
+  command = mach_header + 1;
+  for (command_index = 0; command_index != mach_header->ncmds; command_index++)
+  {
+    const GumLoadCommand * lc = command;
+
+    if (lc->cmd == GUM_LC_DYLD_CHAINED_FIXUPS)
+    {
+      const GumLinkeditDataCommand * fixups = command;
+      const GumChainedFixupsHeader * fixups_header;
+      const GumChainedStartsInImage * image_starts;
+      guint seg_index;
+
+      fixups_header = (const GumChainedFixupsHeader *)
+          ((const guint8 *) image->linkedit + fixups->dataoff);
+
+      image_starts = (const GumChainedStartsInImage *)
+          ((const guint8 *) fixups_header + fixups_header->starts_offset);
+
+      for (seg_index = 0; seg_index != image_starts->seg_count; seg_index++)
+      {
+        const guint seg_offset = image_starts->seg_info_offset[seg_index];
+        const GumChainedStartsInSegment * seg_starts;
+        GumChainedPtrFormat format;
+        guint16 page_index;
+
+        if (seg_offset == 0)
+          continue;
+
+        seg_starts = (const GumChainedStartsInSegment *)
+            ((const guint8 *) image_starts + seg_offset);
+        format = seg_starts->pointer_format;
+
+        segment = gum_find_segment_by_offset (self,
+            seg_starts->segment_offset);
+
+        if (segment == NULL)
+          continue;
+
+        for (page_index = 0; page_index != seg_starts->page_count; page_index++)
+        {
+          guint16 start;
+          const guint8 * cursor;
+
+          start = seg_starts->page_start[page_index];
+          if (start == GUM_CHAINED_PTR_START_NONE)
+            continue;
+
+          cursor = (const guint8 *) mach_header + seg_starts->segment_offset +
+              (page_index * seg_starts->page_size) +
+              start;
+
+          if (format == GUM_CHAINED_PTR_64 ||
+                format == GUM_CHAINED_PTR_64_OFFSET)
```

Comment:
It looks like the second line here should line up with the start of `format` above -- i.e. a hanging indent of 4.